README.md
# pkme
This package is designed for classic compartmental or physiologically-based pharmacokinetic models. These types of models are most frequently described by systems 
of linear ordinary differential equations ODEs. This package takes a matrix-based approach by finding the eigenvalue/eigenvector solution to a linear system of 
ODEs. Matrix manipulation is computationally fast in R, making it ideal for model fitting/parameter estimation, particularly for large models. This function most 
benefits models/systems with high dimensionality and/or problems with stiffness. The manner in which the exact solutions are solved for and integrated allows for 
relatively accurate estimates while using larger step sizes than traditional methods - increasing computation speed. We do not claim that this method is faster than
traditional methods for all models, but it is a nice option for difficult models.  

Structure:
It is designed to take an initial condition matrix (single column vector of initial concentrations) 
and a rate coefficient matrix (flow/volume), dose profile (number of doses, start and stop time of each dose)
along with any parameters needed to solve the system, and outputs a matrix of the concentrations in each 
compartment (rows) versus the time course (columns) for a given step size `h`. Each column represents a 
predicted data point at every `2*h`. 

Additional details:

**Rate Coefficient Matrix** `(rate_coeff)` : your flow per volume and any other parameters and microconstants included in the equations – as you’re looking at the matrix, rows represent compartments and columns represent concentrations

 **Initial Conditions Matrix** `(init_condition)`: just your initial conditions or in this case, initial concentrations in each compartment. Normally a vector of zeroes, but if you did have existing drug concentrations in some organs, it could be represented here
 
**Dose Vector** `(dose)` : where the dose is designated – the nice thing about this function is you can dose into any compartment just by specifying the row in the vector corresponding to that compartment. Let’s say you’re dosing IM and your muscle compartment is #4, you would place your dose in this vector in row 4. Homogenous and non-homogenous linear systems are handled using the same function. For homogenous systems, the "g vector" corresponding to the dose profile can be replaced by setting the total dose as the initial condition for the first compartment and the dose amount in the function to 0.


Also, some first order linear systems produce complex conjugate pairs in the eigenvalues or eigenvectors of the matrix solution. This function handles those complex conjugate pairs automatically.

# Installation Shortcut

Install the latest release on CRAN with:
```
install.packages("pkme")
```

Install the current development version from GitHub with:
```
# install.packages("devtools")
devtools::install_git("git://github.com/hillaryhusband/pkme.git")
```

# Examples:

## One-compartment IV bolus:

```
library(pkme)
```

Model structure:
dC/dt = -k10*C

Designate PK parameters
```
CL = 2.08 #(mL/min)
V = 7.88 #(mL)
k10 = CL/V #(1/min)
# create parameter vector to feed into solver
params <- c(CL, V, k10)
```

Create matrix form of model
```
pmat = matrix(0, nrow=1, ncol=1)
pmat[1,1] = -k10
```

Initialize initial conditions matrix
```
init = matrix(0, nrow=1, ncol=1)
```

Designate time:
```
tfinal = 10  # end of analysis (min)
dt = 0.5     # time step (min)
```

Set-up dose:
```
begin <- 0
end <- dt # for bolus, set infusion rate to length of time step
dosen <- 1
doseamt <- 30
```

Solve system:
```
model1 <- pkme(minute = tfinal,
               h = dt,
               rate_coeff = pmat,
               init_condition = init,
               pars = params,
               number_of_doses = dosen,
               dose = doseamt,
               start_time = begin,
               stop_time = end)
```

Output prints concentration at each time step for each compartment in the system. This code is available in the R directory as `R/pkme-bolus-example`. 

***NOTE: Three-compartment PK model vignette is available in the R directory as `R/pkme-infusion-example.R`.***


NEWS.md
# pkme 0.2.3

# pkme 0.2.2

# pkme 0.2.1

1
# pkme 0.1.0

* Initial CRAN submission.


R/pkme.R
#' Solve PK or PBPK model in matrix form
#'
#' Takes in initial condition matrix (single column vector of initial concentrations) and a rate
#' coefficient matrix (flow/volume), dose profile (number of doses, start and stop time of each dose)
#' along with any parameters needed to solve the system, and outputs a matrix of the concentrations
#' in each compartment (rows) versus the time course (columns) for a given step size "h". Each column
#' represents a predicted data point at every "2*h"
#' @param minute The number of minutes the model will simulate out
#' @param h The step size specified for Simpson's 1/3 rule
#' @param rate_coeff A matrix of rates - flow/volume corresponding to the equation/compartment (row) and concentration to be multiplied by (column)
#' @param init_condition A column vector of initial conditions (concentrations) for each equation/compartment (row)
#' @param pars A matrix of parameters defining the variables in the rate coefficient matrix and the initial conditions, if necessary
#' @param number_of_doses A scalar indicating number of doses
#' @param dose A vector of doses corresponding to the "ith" start and stop time vector
#' @param start_time A vector of start times for an "i" number of doses
#' @param stop_time A vector of stop times for an "i" number of doses
#' @param dose_comp Designate which compartment to place the dose -> default is 1
#' @examples
#' # Simple one-compartment PK model:
#' dt = 0.5
#' pmat = matrix(0, nrow=1, ncol=1)
#' pmat[1,1] = -k10
#' init = matrix(0, nrow=1, ncol=1)
#' pkme(minute = 10, h = dt, rate_coeff = pmat, init_condition = init,
#'             pars = c(CL = 2.08, V = 7.88, k10 = CL/V), number_of_doses = 1,
#'             dose = 30, start_time = 0, stop_time = dt)
#'             #(for bolus, set length of infusion equal to 'dt'/time step)
#' @return x_matrix_save A matrix of the concentrations in each compartment (rows) versus the time course (columns) for a given step size "h". Each column represents a predicted data point at every "2*h"
#' @export


pkme <- function(minute, h, rate_coeff, init_condition, pars, number_of_doses, dose, start_time, stop_time, dose_comp = 1) {
  t_index_max = (minute / h) + 1
  tol_value = 10^-500
  matrix_size = nrow(rate_coeff)
  injection_rate <- matrix(0, nrow = t_index_max, ncol=1)
  for (t_index in 1:t_index_max) {
   t <- (t_index - 1) * h
   for (i in 1:number_of_doses) {
     if (t <= stop_time[i] && t > start_time[i]) {injection_rate[t_index,] <- dose[i] / (stop_time[i] - start_time[i])}
   }
  }


  if (nrow(rate_coeff) != ncol(rate_coeff)) {
    stop('matrix must be square')
  }

  if (nrow(rate_coeff) != nrow(init_condition)) {
    stop('initial conditions must be same size as rate coefficient matrix')
  }
  # eigenvalue/eigenvector solution

  # eigenvalues and eigenvectors of coefficient matrix
  eigen_matrix <- eigen(rate_coeff)

  # print eigenvalues and eigenvectors
  eigen_matrix$values
  eigen_matrix$vectors

  # separate real and imaginary parts of eigenvectors
  a_real_eigenvec <- Re(eigen_matrix$vectors)
  b_imag_eigenvec <- Im(eigen_matrix$vectors)

  # print real and complex eigenvectors
  a_real_eigenvec
  b_imag_eigenvec

  # separate real and imaginary parts of eigenvalues
  lambda <- Re(eigen_matrix$values)
  mu <- Im(eigen_matrix$values)

  # print real and complex eigenvalues
  lambda
  mu

  # initialize matrices
  conc_v_time <- vector()
  u_matrix <- matrix(0, nrow = matrix_size, ncol = 1)
  g_vector <- matrix(0, nrow = matrix_size, ncol = 1)
  u_prime_matrix <- matrix()
  u_prime_matrix_save <- matrix(0, nrow = matrix_size, ncol = 1)
  fund_matrix <- matrix(0, nrow = matrix_size, ncol = matrix_size)

  t = 0
  ## calculate fundamental matrix and inverse of fundamental matrix for 0, h, 2*h
  while (t <= h*2) {
    i <- 1
    while (i <= matrix_size){
      j <- 1
      while (j <= matrix_size) {
        if (mu[j] == 0) {
          fund_matrix[i,j] <- exp(lambda[j]*t)*(a_real_eigenvec[i,j] * cos(mu[j]*t))
          j <- j + 1
        }
        else {
          fund_matrix[i,j] <- exp(lambda[j]*t) * (a_real_eigenvec[i,j] * cos(mu[j]*t) - b_imag_eigenvec[i,j] * sin(mu[j]*t))
          j <- j + 1
          fund_matrix[i,j] <- exp(lambda[j]*t) * (a_real_eigenvec[i,j] * sin(mu[j]*t) + b_imag_eigenvec[i,j] * cos(mu[j]*t))
          j <- j + 1
        }
      }
      i = i + 1
    }

    if (t == 0) {
      t0_fund_matrix <- fund_matrix
    } else if (t == h * 1) {
      t1_fund_matrix <- fund_matrix
    } else if (t == h * 2 ) {
      t2_fund_matrix <- fund_matrix
    } else {
      fund_matrix <- fund_matrix
    }

    t = t + h
  }

  t0_inv_fund_matrix = solve(t0_fund_matrix, tol = tol_value)
  t1_inv_fund_matrix = solve(t1_fund_matrix, tol = tol_value)
  t2_inv_fund_matrix = solve(t2_fund_matrix, tol = tol_value)

  ## function starts here

  for (t_index in seq(from = 1, to = t_index_max, by = 1)) {
    g_vector[dose_comp,1] <- injection_rate[t_index]

    u_prime_matrix0 <- t0_inv_fund_matrix %*% g_vector
    u_prime_matrix1 <- t1_inv_fund_matrix %*% g_vector
    u_prime_matrix2 <- t2_inv_fund_matrix %*% g_vector

    ## initial time,
    ## run once to get constant vector
    t0_inv_fund_matrix = solve(t0_fund_matrix, tol = tol_value)
    # constants for a given initial condition at t = 0
    int_constants =   t0_inv_fund_matrix %*% init_condition

    ## Simpson's 1/3 rule for 0 to 0.2
    sim_soln <- (h/3) * (u_prime_matrix0 + (4 * u_prime_matrix1) + u_prime_matrix2)
    sim_soln <- matrix(sim_soln)

    ## u_matrix should be term by term u_prime_matrix integrated w.r.t. t and bounds 0 to time point
    x_matrix = fund_matrix %*% int_constants + fund_matrix %*% sim_soln
    conc_v_time <- cbind(conc_v_time, x_matrix)

    ## x_matrix solution becomes initial condition for next tau
    init_condition = x_matrix
   }
   return(conc_v_time)
  }


pbpkme.R
#' Solve PBPK model in matrix form
#'
#' Takes in initial condition matrix (single column vector of initial concentrations) and a rate
#' coefficient matrix (flow/volume), dose profile (number of doses, start and stop time of each dose)
#' along with any parameters needed to solve the system, and outputs a matrix of the concentrations
#' in each compartment (rows) versus the time course (columns) for a given step size "h". Each column
#' represents a predicted data point at every "2*h"
#' @param minute The number of minutes the model will simulate out
#' @param h The step size specified for Simpson's 1/3 rule
#' @param rate_coeff A matrix of rates - flow/volume corresponding to the equation/compartment (row) and concentration to be multiplied by (column)
#' @param init_condition A column vector of initial conditions (concentrations) for each equation/compartment (row)
#' @param pars A matrix of parameters defining the variables in the rate coefficient matrix and the initial conditions, if necessary
#' @param number_of_doses A scalar indicating number of doses
#' @param dose A vector of doses corresponding to the "ith" start and stop time vector
#' @param start_time A vector of start times for an "i" number of doses
#' @param stop_time A vector of stop times for an "i" number of doses
#' @param dose_comp Designate which compartment to place the dose -> default is 1
#' @return x_matrix_save A matrix of the concentrations in each compartment (rows) versus the time course (columns) for a given step size "h". Each column represents a predicted data point at every "2*h"
#' @export


pbpkme <- function(minute, h, rate_coeff, init_condition, pars, number_of_doses, dose, start_time, stop_time, dose_comp = 1) {
  t_index_max = (minute / h) + 1
  tol_value = 10^-500
  matrix_size = nrow(rate_coeff)
  injection_rate <- matrix(0, nrow = t_index_max, ncol=1)
  for (t_index in 1:t_index_max) {
   t <- (t_index - 1) * h
   for (i in 1:number_of_doses) {
     if (t <= stop_time[i] && t > start_time[i]) {injection_rate[t_index,] <- dose[i] / (stop_time[i] - start_time[i])}
   }
  }


  if (nrow(rate_coeff) != ncol(rate_coeff)) {
    stop('matrix must be square')
  }

  if (nrow(rate_coeff) != nrow(init_condition)) {
    stop('initial conditions must be same size as rate coefficient matrix')
  }
  # eigenvalue/eigenvector solution

  # eigenvalues and eigenvectors of coefficient matrix
  eigen_matrix <- eigen(rate_coeff)

  # print eigenvalues and eigenvectors
  eigen_matrix$values
  eigen_matrix$vectors

  # separate real and imaginary parts of eigenvectors
  a_real_eigenvec <- Re(eigen_matrix$vectors)
  b_imag_eigenvec <- Im(eigen_matrix$vectors)

  # print real and complex eigenvectors
  a_real_eigenvec
  b_imag_eigenvec

  # separate real and imaginary parts of eigenvalues
  lambda <- Re(eigen_matrix$values)
  mu <- Im(eigen_matrix$values)

  # print real and complex eigenvalues
  lambda
  mu

  # initialize matrices
  conc_v_time <- vector()
  u_matrix <- matrix(0, nrow = matrix_size, ncol = 1)
  g_vector <- matrix(0, nrow = matrix_size, ncol = 1)
  u_prime_matrix <- matrix()
  u_prime_matrix_save <- matrix(0, nrow = matrix_size, ncol = 1)
  fund_matrix <- matrix(0, nrow = matrix_size, ncol = matrix_size)

  t = 0
  ## calculate fundamental matrix and inverse of fundamental matrix for 0, h, 2*h
  while (t <= h*2) {
    i <- 1
    while (i <= matrix_size){
      j <- 1
      while (j <= matrix_size) {
        if (mu[j] == 0) {
          fund_matrix[i,j] <- exp(lambda[j]*t)*(a_real_eigenvec[i,j] * cos(mu[j]*t))
          j <- j + 1
        }
        else {
          fund_matrix[i,j] <- exp(lambda[j]*t) * (a_real_eigenvec[i,j] * cos(mu[j]*t) - b_imag_eigenvec[i,j] * sin(mu[j]*t))
          j <- j + 1
          fund_matrix[i,j] <- exp(lambda[j]*t) * (a_real_eigenvec[i,j] * sin(mu[j]*t) + b_imag_eigenvec[i,j] * cos(mu[j]*t))
          j <- j + 1
        }
      }
      i = i + 1
    }

    if (t == 0) {
      t0_fund_matrix <- fund_matrix
    } else if (t == h * 1) {
      t1_fund_matrix <- fund_matrix
    } else if (t == h * 2 ) {
      t2_fund_matrix <- fund_matrix
    } else {
      fund_matrix <- fund_matrix
    }

    t = t + h
  }

  t0_inv_fund_matrix = solve(t0_fund_matrix, tol = tol_value)
  t1_inv_fund_matrix = solve(t1_fund_matrix, tol = tol_value)
  t2_inv_fund_matrix = solve(t2_fund_matrix, tol = tol_value)

  ## function starts here

  for (t_index in seq(from = 1, to = t_index_max, by = 1)) {
    g_vector[dose_comp,1] <- injection_rate[t_index]

    u_prime_matrix0 <- t0_inv_fund_matrix %*% g_vector
    u_prime_matrix1 <- t1_inv_fund_matrix %*% g_vector
    u_prime_matrix2 <- t2_inv_fund_matrix %*% g_vector

    ## initial time,
    ## run once to get constant vector
    t0_inv_fund_matrix = solve(t0_fund_matrix, tol = tol_value)
    # constants for a given initial condition at t = 0
    int_constants =   t0_inv_fund_matrix %*% init_condition

    ## Simpson's 1/3 rule for 0 to 0.2
    sim_soln <- (h/3) * (u_prime_matrix0 + (4 * u_prime_matrix1) + u_prime_matrix2)
    sim_soln <- matrix(sim_soln)

    ## u_matrix should be term by term u_prime_matrix integrated w.r.t. t and bounds 0 to time point
    x_matrix = fund_matrix %*% int_constants + fund_matrix %*% sim_soln
    conc_v_time <- cbind(conc_v_time, x_matrix)

    ## x_matrix solution becomes initial condition for next tau
    init_condition = x_matrix
   }
   return(conc_v_time)
  }


R/pkme-bolus-example.R
#library(pkme)
source("R/pkme.R")
# Model structure:

# dC/dt = -k10*C

# Designate PK parameters
CL = 2.08 #(mL/min)
V = 7.88 #(mL)
k10 = CL/V #(1/min)

params <- c(CL, V, k10)

# Create matrix form of model

pmat = matrix(0, nrow=1, ncol=1)
pmat[1,1] = -k10

# Initialize initial conditions matrix

init = matrix(0, nrow=1, ncol=1)

# Designate time:
tfinal = 10  # end of analysis (min)
dt = 0.5     # time step (min)


# Set-up dose:
begin <- 0
end <- dt # for bolus, set infusion rate to length of time step
dosen <- 1
doseamt <- 30


# Solve system:

model1 <- pkme(minute = tfinal,
               h = dt,
               rate_coeff = pmat,
               init_condition = init,
               pars = params,
               number_of_doses = dosen,
               dose = doseamt,
               start_time = begin,
               stop_time = end)


R/pkme-infusion-example.R
rm(list=ls())

## run this chunk once to install package needed to download 'pbpkme' from Github
## skip if you've already done this
install.packages("devtools")
# library(devtools)
# install_github("hillaryhusband/pkme")
############
# library(pkme)
source("R/pkme.R")
library(pracma)
library(readxl)
library(lattice)
library(ggplot2)
library(tidyverse)
library(ggthemes)

dt = 0.01
tfinal = 60

### Parameters and model based on Remifentanil model by Cascone et al.:
##  Cascone S, Lamberti G, Titomanlio G, Piazza O. Pharmacokinetics of Remifentanil:
## a three-compartmental modeling approach. Translational Medicine. 2013;7(4):18–22

#parameters
# k10 <- 0.172   min^-1
# k12 <- 0.373   min^-1
# k13 <- 0.0367  min^-1
# k21 <- 0.103   min^-1
# k31 <- 0.0124  min^-1
#
# V1 <- 7.88 mL
# V2 <- 23.9 mL
# V3 <- 13.8 mL
#
# CL1 <- 2.08    mL/min
# CL2 <- 0.828   mL/min
# CL3 <- 0.0784  mL/min

body_weight <- 79 # kg


params <- c(k10 = 0.172, k12 = 0.373, k13 = 0.0367, k21 = 0.103, k31 = 0.0124, V1 = 7.88,
            V2 = 23.9, V3 = 13.8, CL1 = 2.08, CL2 = 0.828, CL3 = 0.0784)
# drug injection
# R_in = 5 * body_weight

# C1 = 0
# C2 = 0
# C3 = 0

init_cond_matrix = matrix(0, nrow = 3, ncol = 1)

init_cond_matrix[1,1] = 0 #C1
init_cond_matrix[2,1] = 0 #C2
init_cond_matrix[3,1] = 0 #C3

#dC1_dt = (((-CL1*C1 + k21*V2*C2 + k31*V3*C3 - (k12+k13+k10)*C1)*V1) + R_in) / V1
#dC2_dt = (k12 *C1*V1 - k21*C2*V2 - CL2*C2) / V2
#dC3_dt = (k13*C1*V1 - k31*C3*V3 - CL3*C3) / V3


p_matrix = matrix(0, nrow=3, ncol=3)
p_matrix[1,1] = -((CL1/V1) + k12 + k13 + k10)
p_matrix[1,2] = (k21 * V2) / V1
p_matrix[1,3] = (k31 * V3) / V1
p_matrix[2,1] = (k12 * V1) / V2
p_matrix[2,2] = -((CL2/V2) + k21)
p_matrix[3,1] = (k13 * V1) / V3
p_matrix[3,3] = -((CL3/V3) + k31)


begin <- 0
end <- 5
dose_number <- 1
dose_amt <- 30


model1 <- pkme(minute = tfinal, h = dt, rate_coeff = p_matrix, init_condition = init_cond_matrix,
                 pars = params, number_of_doses = dose_number, dose = dose_amt, start_time = begin,
                 stop_time = end)

## create continuous time vector and pull out compartments from final matrix
time <- seq(from=0, to = tfinal, by = dt)
conc1 <- model1[1,]
conc2 <- model1[2,]
conc3 <- model1[3,]


## merge time vector and concentrations for plotting in ggplot
C1 <- data.frame(time, conc1)
C2 <- data.frame(time, conc2)
C3 <- data.frame(time, conc3)


plot1 <- ggplot(data = C1, aes(x = time, y = conc1))+ geom_line() + theme_base() +
   xlab("Time (minutes)") + ylab("Concentration (microMolar)") + ggtitle("Concentration vs Time", subtitle = "Compartment 1") + scale_y_log10()
plot2 <- ggplot(data = C1, aes(x = time, y = conc2))+ geom_line() + theme_base() +
   xlab("Time (minutes)") + ylab("Concentration (microMolar)") + ggtitle("Concentration vs Time", subtitle = "Compartment 2") + scale_y_log10()
plot3 <- ggplot(data = C1, aes(x = time, y = conc3))+ geom_line() + theme_base() +
   xlab("Time (minutes)") + ylab("Concentration (microMolar)") + ggtitle("Concentration vs Time", subtitle = "Compartment 3") + scale_y_log10()


R/pbpkme_examplescript.R
rm(list=ls())

## run this chunk once to install package needed to download 'pbpkme' from Github
## skip if you've already done this
install.packages("devtools")
library(devtools)
install_github("hillaryhusband/pbpkme")
############
library("pbpkme")
library(pracma)
library(readxl)
library(lattice)
library(ggplot2)
library(tidyverse)
library(ggthemes)

dt = 0.01
tfinal = 60

### Parameters and model based on Remifentanil model by Cascone et al.:
##  Cascone S, Lamberti G, Titomanlio G, Piazza O. Pharmacokinetics of Remifentanil:
## a three-compartmental modeling approach. Translational Medicine. 2013;7(4):18–22

#parameters
k10 <- 0.172    # min^-1
k12 <- 0.373    # min^-1
k13 <- 0.0367   # min^-1
k21 <- 0.103    # min^-1
k31 <- 0.0124   # min^-1

V1 <- 7.88 #mL
V2 <- 23.9 #mL
V3 <- 13.8 #mL

CL1 <- 2.08   # mL/min
CL2 <- 0.828  # mL/min
CL3 <- 0.0784 # mL/min

body_weight <- 79 # kg (mean value)

p_matrix = matrix(0, nrow=3, ncol=3)
p_matrix[1,1] = -((CL1/V1) + k12 + k13 + k10)
p_matrix[1,2] = (k21 * V2) / V1
p_matrix[1,3] = (k31 * V3) / V1
p_matrix[2,1] = (k12 * V1) / V2
p_matrix[2,2] = -((CL2/V2) + k21)
p_matrix[3,1] = (k13 * V1) / V3
p_matrix[3,3] = -((CL3/V3) + k31)


params <- c(k10 = 0.172, k12 = 0.373, k13 = 0.0367, k21 = 0.103, k31 = 0.0124, V1 = 7.88,
            V2 = 23.9, V3 = 13.8, CL1 = 2.08, CL2 = 0.828, CL3 = 0.0784)
# drug injection
# R_in = 5 * body_weight

# C1 = 0
# C2 = 0
# C3 = 0

init_cond_matrix = matrix(0, nrow = 3, ncol = 1)

init_cond_matrix[1,1] = 0 #C1
init_cond_matrix[2,1] = 0 #C2
init_cond_matrix[3,1] = 0 #C3

#dC1_dt = (((-CL1*C1 + k21*V2*C2 + k31*V3*C3 - (k12+k13+k10)*C1)*V1) + R_in) / V1
#dC2_dt = (k12 *C1*V1 - k21*C2*V2 - CL2*C2) / V2
#dC3_dt = (k13*C1*V1 - k31*C3*V3 - CL3*C3) / V3


p_matrix = matrix(0, nrow=3, ncol=3)
p_matrix[1,1] = -((CL1/V1) + k12 + k13 + k10)
p_matrix[1,2] = (k21 * V2) / V1
p_matrix[1,3] = (k31 * V3) / V1
p_matrix[2,1] = (k12 * V1) / V2
p_matrix[2,2] = -((CL2/V2) + k21)
p_matrix[3,1] = (k13 * V1) / V3
p_matrix[3,3] = -((CL3/V3) + k31)


begin <- 0
end <- 0.5
dose_number <- 1
dose_amt <- 30


model1 <- pbpkme(minute = tfinal, h = dt, rate_coeff = p_matrix, init_condition = init_cond_matrix,
                 pars = params, number_of_doses = dose_number, dose = dose_amt, start_time = begin,
                 stop_time = end)

## create continuous time vector and pull out compartments from final matrix
time <- seq(from=0, to = tfinal, by = dt)
conc1 <- model1[1,]
conc2 <- model1[2,]
conc3 <- model1[3,]


## merge time vector and concentrations for plotting in ggplot
C1 <- data.frame(time, conc1)
C2 <- data.frame(time, conc2)
C3 <- data.frame(time, conc3)


plot1 <- ggplot(data = C1, aes(x = time, y = conc1))+ geom_line() + theme_base() +
   xlab("Time (minutes)") + ylab("Concentration (microMolar)") + ggtitle("Concentration vs Time", subtitle = "Compartment 1") + scale_y_log10()
plot2 <- ggplot(data = C1, aes(x = time, y = conc2))+ geom_line() + theme_base() +
   xlab("Time (minutes)") + ylab("Concentration (microMolar)") + ggtitle("Concentration vs Time", subtitle = "Compartment 2") + scale_y_log10()
plot3 <- ggplot(data = C1, aes(x = time, y = conc3))+ geom_line() + theme_base() +
   xlab("Time (minutes)") + ylab("Concentration (microMolar)") + ggtitle("Concentration vs Time", subtitle = "Compartment 3") + scale_y_log10()
